### PR TITLE
InheritedElement.removeDependent()

### LINKED ISSUE #129207: There's no way for InheritedElement to know that a dependent has been deactivated
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

I'm finishing the promising-looking, easy-to-use approach to state management, which requires that `InheritedElement` is notified when the dependent has been removed from the dependents list to properly close all value change subscriptions upon dependent element deactivation.

### Proposal

Call the `dependency.onDependentRemoved(this)` right after `dependency._dependents.remove(this)` inside the `Element.deactivate()`

### PR DESCRIPTION
**Pull request by s0nerik**
Call the `dependency.removeDependent(this)` instead of `dependency._dependents.remove(this)` inside the `Element.deactivate()`. This allows `InheritedElements` to know when they can release resources associated with a given dependent `Element`.

Fixes #129207 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by s0nerik** [comment]
@goderbauer I've added the test. Please take a look.

**Comment by goderbauer** [comment]
There's a lengthy discussion about exactly this API in https://github.com/flutter/flutter/issues/128432, which hasn't reached a conclusion yet and we should hold off on this change until it does. I would encourage you to post your use case there as that seems to be what's still missing: a compelling argument for actually having this API.

**Comment by s0nerik** [comment]
I already have a slightly adjusted version of this PR that actually solves #128432. @goderbauer please let me know if I should create a new PR for that or just push these changes into this one and adjust the description a bit.

**Comment by goderbauer** [comment]
@s0nerik I lost track of this PR and I apologize for the delay. If we actually decide to do this, I would prefer the approach that was outlined in the first few comments of https://github.com/flutter/flutter/issues/128432 (i.e. introduce a removeDependency method to complete the existing set of update, get, set methods) instead of adding a separate notification mechanism as implemented in this PR. Is that the change you had in mind with your previous comment? Would you mind pushing that into this PR so we can evaluate that?

**Comment by s0nerik** [comment]
Hey @goderbauer, thanks for reaching out. I've pushed the changes I mentioned before along with some doc comment adjustments for clarity.

BTW, I named the method `removeDependent`, not `removeDependency` since it makes much more sense to me that way as it is defined on an `InheritedElement` and is supposed to remove another `Element` from the list of dependents of this `InheritedElement`.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:6028]:
```suggestion
  /// Called by [Element.deactivate] to remove the provided `dependent` [Element] from this [InheritedElement].
  ///
  /// After the dependent is removed, [Element.didChangeDependencies] will no
  /// longer be called on it when this [InheritedElement] notifies its dependents.
```

**Review by goderbauer — COMMENTED** [review]
LGTM after doc nit is resolved.

**Comment by goderbauer** [comment]
@s0nerik Can you rebase this with the latest master to make the checks happy?

**Comment by s0nerik** [comment]
@goderbauer I did. Please take a look.

**Comment by s0nerik** [comment]
@goderbauer All checks seem to be happy now.

**Review by goderbauer — APPROVED** [review]
LGTM

**Review by LongCatIsLooong — APPROVED** [review]

**Pull request merged**
